### ISSUE TITLE
Implement Isloate.spawnSync

### ISSUE THREAD

**Issue by bsutton** [dart-lang/sdk#55356]
I'm the author of DCli which utilises waitFor to synchronously spawn processes.

With the deprecation of waitFor I'm currently stuck without a way to maintain the current api.

This issue tracking the deprecation of waitFor is here:

https://github.com/dart-lang/sdk/issues/52121

The core issue is that I need to spawn an isolate synchronously.

I'm not looking for the isolate to run to completion, rather just the ability to start the isolate synchronously.

With the current api
```dart
final isolate = Isolate.spawn<Args>(_runProcess, Args(commandToRun, mailbox));

/// The following code creates a dead lock as the isolate hasn't started and so doesn't
/// put anything into the mailbox.
mailbox.take(); // sync call that doesn't return util the isolate starts and puts a message into the mailbox
```

```dart
final isolate = Isolate.spawnSync<Args>(_runProcess, Args(commandToRun, mailbox));
/// I'm now guaranteed the isolate is running which means it will eventually place something in the mailbox.
mailbox.take(); // sync call.
```

I've tried an number of paths to get around the spawn issue but it appears that if you make a sync call (e.g. mailbox.take) immediately after calling Isolate.spawn then the spawned isolate doesn't start. I assume the current spawn logic is relying on a async call to be given processing time to start.

Unless I'm wrong about how spawn works, I've hit a brick wall in migrating dcli off of waitfor. I have resolutions for all of the other code except this rather critical piece.

@mraleph 
@Hixie 
@itsjustkevin 

I need this enhancement before waitFor is removed from the dart api.

**Comment by tsavo-at-pieces** [comment]
Good looks! +1

**Comment by nex3** [comment]
I agree that this is likely to be necessary for moving off _all_ uses of `waitFor()`.

**Comment by bsutton** [comment]
I've would like to see waitfor un-deproacated until this is in the stable
release.

I have everything else ready but dcli is broken against stable until we
have something like this as a solution.

**Comment by lrhn** [comment]
> I assume the current spawn logic is relying on a async call to be given processing time to start.

That _should_ not be necessary.
At the time the future is returned, the spawn should be underway, with no more local-isolate work necessary. The local future won't be completed while the local isolate is blocked, but if you don't care about that, it shouldn't be a problem.

If it doesn't work that way, I'll consider it a bug.

Could it be a problem that you're receiving events on other ports, which cannot be processed, and that somehow blocks the event to the port you're using for the mailbox? (If so, that should probably also be considered a bug, if nowhere else then in the mailbox code.)

**Comment by lrhn** [comment]
Actually, the problem is probably that the spawn *fails*, which is reported as an asynchronous exception in the returned future. Which means it's not reported when the isolate is blocked, but since the isolate never started, the mailbox won't get any messages to unlock it. 
That's a deadlock.

A reason it could throw might be that you're sending the mailbox object directly, and it's not sendable.
Change the code to:
```dart
final isolate = Isolate.spawn<Args>(_runProcess, Args(commandToRun, mailbox.asSendable));
```
and unwrap the `Sendable<MailBox>` using `.materialize()` at the other end.

Or, what I would do, make the `Args` constructor call `.asSendable` on its argument and store it as a `Sendable<MailBox>`. The give it a `mailbox` getter which re-materializes the mailbox if needed. If you want to be absolutely certain that the mailbox isn't stored on the object, in case you want to send it to another isolate, use an `Expando`.

Something like:
```dart
class Args {
  static Expando<Maibox> _mailboxCache = Expando();
  final Sendable<Mailbox> _mailboxPackage;
  final Command commandToRun;
  Args(this.commandToRun, MailBox mailbox) : _mailboxPackage = mailbox.asSendable {
    _mailboxCache[this] = mailbox;
  }
  Mailbox get mailbox => _mailboxCache ??= _mailboxPackage.materialize();
}
```
That should contain the sending logic, while still letting both ends use the object like today.

It *is* a problem that isolate spawn can fail *later*, but that's what asynchronous computations do.
A synchronous spawn would prevent that. Alternatively, what is needed is not a synchronous spawn, but just synchronous validation of the message. If a flag to `Isolate.spawn` could force synchronous validation of the sendability of the function and argument, and throw a synchronous `ArgumentError` (or whatever) if not sendable, then the actual spawning (which should then not fail for any reason other than out-of-memory) could be asynchronous.

Or maybe `mailbox.take` needs a timeout. If no message has been received in time, it throws a `TimeoutException` instead, which allows reporting an error instead of dead-locking. Tricky, if we're doing arbitrary user requested operations. Maybe we can have the operations send progress reports in the mailbox, like:
```dart
var isolateFuture = Isolate.spawn(_run, _Args(data, mailbox));
try {
  Uint8List msg;
  do {
    msg = mailbox.take(timeout: const Duration(seconds: 5);
  } while (_isStillWorking(msg)); // Standard IAmAlive report sent every second or so, not final result.
  return _parseResult(msg);
} on TimeoutException {
  isolateFuture.then((isolate) {
    isolate.kill();
  });
  throw OperationFailedException("User operation stalled");
}
```
Then you could build a protocol on top of the mailbox, instead of only using it for the final result.
(But without a time-out, it's limited how that can be used to prevent deadlocks.)

**Comment by tsavo-at-pieces** [comment]
@lrhn  I was thinking the same and asked that as well in the other issue thread: https://github.com/dart-lang/sdk/issues/52121#issuecomment-2033107323

@bsutton are you sure this is not the case?

**Comment by tsavo-at-pieces** [comment]
Just kidding - this was the code that I updated to use native_synchronization

i.e. 

```
SendPort _connectSendPort(ProcessChannel channel) {
  /// take the initial message which contains
  /// the channels sendPort id.
  final msg = channel.send.take();

  return NativeCalls.connectToPort(msg);
}
```
[from this file](https://github.com/onepub-dev/dcli/blob/2e245773ca1fb5b067f6b121d390a8b2baeb499a/dcli/lib/src/process/process/process_in_isolate.dart#L34C3-L34C35)

is referencing: 

```dart 
class ProcessChannel {
  ProcessChannel()
      : pipeMode = false,
        response = Mailbox(),
        send = Mailbox();
 ```
[in this file](https://github.com/onepub-dev/dcli/blob/2e245773ca1fb5b067f6b121d390a8b2baeb499a/dcli/lib/src/process/process/process_channel.dart#L23)

And I forgot that I had updated this strong type to verify here:
```dart 
void _startIsolate(ProcessSettings processSettings, ProcessChannel channel) {
unawaited(Isolate.spawn<List<Sendable<Mailbox>>>((mailboxes) async {
```

at [this line](https://github.com/onepub-dev/dcli/blob/2e245773ca1fb5b067f6b121d390a8b2baeb499a/dcli/lib/src/process/process/process_in_isolate.dart#L41)

**Comment by bsutton** [comment]
Given the above feedback i decided to write a few simplified experiments.

The following works as expected:

```dart
void main() {
  unawaited(Isolate.spawn(_run, 'hellow'));
  sleep(const Duration(seconds: 10));
}

void _run(String arg) {
  print('Hello from isolate: $arg');
}
```
This seems to support @lrhn statement that the spawn should proceed even if a sync call is made immediately after the spawn occurs.

**Comment by bsutton** [comment]
Doing a simple mailbox.take also allows the isolate to spawn without issue:

```dart
void test2() {
  final mailbox = Mailbox();
  unawaited(Isolate.spawn(_run, 'hellow'));
  mailbox.take();
}

void _run(String arg) {
  print('Hello from isolate: $arg');
}

```

**Comment by lrhn** [comment]
> @lrhn Maybe you could triple spot check me here just to make sure I didn't overlook something obvious?

Not going to try to grok all the code, and don't know enough about the native synchronization functions, so I'll just give the completely generic recommendation to avoid the risk of over-capturing with closures.

The `(mailboxes) async { ... }` closure captures the `processSetting` variable from the surrounding scope, and is passed a `List<Sendable<Mailbox>>` as argument. The `channel` containing the original unsendable `Mailbox`es is also in scope. If the closure decides to (over-)capture that variable too, it will be unsendable.
It *probably* won't, because there is no closure capturing `channel` at all. It's still something that can go wrong if the compiler makes a wrong deduction somewhere.

I'd still consider moving the closure to be its own static function, and pass the `processSettings` as an argument too, using a record for the argument and patterns to destructure, so something like:
```dart
void _runInIsolate((ProcessSettings, Sendable<MailBox>, Sendable<Mailbox>) args) {
  final (processSettings, sendableSendBox, sendableResponseBox) = args;
  final sendBox = sendableSendBox.materialize();
  final responseBox = sendableResponseBox.materialize();
  // ...
}
```

Or pack up the mailboxes before calling `_startIsolate(ProcessSettings processSettings, List<Sendable<Mailbox>> mailboxes) {... }` so the unsendable mailboxes are not in the scope at all.

Whether the following logic is correct, that's for tests to decide :grin:.

**Comment by bsutton** [comment]
I've made progress so thanks for the input. I will update in a couple of
days once I get further down the path.

**Comment by bsutton** [comment]
So my statement about how spawn works was wrong, it will launch the process
even if a sync call is made immediate after.

So I have no immediate need for it.

It would still be nice as it would make some scenarios easier to code.
